### PR TITLE
Save parameters as a dictionary & `load_parameters`

### LINKED ISSUE #113: Loading issues when altering policies code structure
**Describe the bug**
When saving, the code only saves in the order each node was created, and not by name. As such, any reordering of creation of the nodes for the policy, will cause unexpected behavior. (found by erniejunior #111)

**Code example**
```python
from stable_baselines import PPO2

PPO2("MlpPolicy", "CartPole-v1").learn(1000).save("model")
```
then change the policy from this:
```python
processed_x = tf.layers.flatten(self.processed_x)
pi_h = processed_x
vf_h = processed_x
for i, layer_size in enumerate(layers):
    pi_h = activ(linear(pi_h, 'pi_fc' + str(i), n_hidden=layer_size, init_scale=np.sqrt(2)))
    vf_h = activ(linear(vf_h, 'vf_fc' + str(i), n_hidden=layer_size, init_scale=np.sqrt(2)))
value_fn = linear(vf_h, 'vf', 1)
pi_latent = pi_h
vf_latent = vf_h
```
to this 
```python
processed_x = tf.layers.flatten(self.processed_x)
pi_h = processed_x
vf_h = processed_x
for i, layer_size in enumerate(layers):
    pi_h = activ(linear(pi_h, 'pi_fc' + str(i), n_hidden=layer_size, init_scale=np.sqrt(2)))
for i, layer_size in enumerate(layers):
    vf_h = activ(linear(vf_h, 'vf_fc' + str(i), n_hidden=layer_size, init_scale=np.sqrt(2)))
value_fn = linear(vf_h, 'vf', 1)
pi_latent = pi_h
vf_latent = vf_h
```
will cause loading to crash
```python
from stable_baselines import PPO2

PPO2.load("model")
```

**System Info**
Describe the characteristic of your environment:
 * installed from source
 * Python 3.6
 * Tensorflow 1.5.0

**Additional context**
the loading and saving should be done by node name.

### LINKED ISSUE #312: [Feature Proposal] `load(...)` but only for model parameters
**Problem**
Currently `ActorCriticRLModel` implements `load(...)` so that it creates the network/model from zero and then loads stored parameters. This creates hefty amount of overhead, and bit of a memory leak if you are not careful, when you only want to change to different set of model parameters, e.g. when evaluating bunch of stored model.

**Suggestion**
* New function for `ActorCriticRLModel` named `load_parameters(str:filename, ...)` which takes in path to a saved model (from `save(...)`) and only loads the model parameters for the agent. Most of this can be done by re-using code already in `load(...)`, with some re-structuring. 
* Required TF `assign` operations would be created during creation of the graph, after `setup_model`. 

We have been using hacky-ish setup like this for our work with success, and thought it could be useful for others as well.

### PR DESCRIPTION
**Pull request by Miffyli**
As discussed in #312, this PR adds two things:

- Model parameters are stored in a dictionary rather than a list. Dictionary keys are tensorflow Variable names.
- Adds `load_parameters` and `get_parameters` functions for loading/getting parameters from an existing model (including tests for these functions).

A modification to `BaseRLModel`: 

- Adds abstract method `_get_parameter_list`, which should return list of tensorflow Variables needed for saving/loading the model. Replaces the lists previously defined in `save`.

closes #312
closes #113

### CODE REVIEW

File: stable_baselines/common/base_class.py [hunk 1]
```diff
@@ -153,6 +155,49 @@ def _setup_learn(self, seed):
         if seed is not None:
             set_global_seeds(seed)
 
+    @abstractmethod
+    def _get_parameter_list(self):
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: quick question: why not a public method? Because it is only used internally?

Comment (Miffyli) [reply stable_baselines/common/base_class.py]: Yes. The intuition was that you could save/load parameters without fiddling with tensorflow stuff. Granted, there shouldn't be anything wrong with accessing this function.

File: stable_baselines/common/base_class.py [hunk 2]
```diff
@@ -153,6 +155,49 @@ def _setup_learn(self, seed):
...
+
+        :return: (list) List of tensorflow Variables
+        """
+        pass
+
+    def get_parameters(self):
+        """
+        Get current model parameters as dictionary of variable name -> ndarray.
+
+        :return: (dict) Dictionary of variable name -> ndarray of model's parameters.
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: Why not an ordered dict?

Comment (Miffyli) [reply stable_baselines/common/base_class.py]: There is no reason to have it ordered apart from having consistency with the previous "lists in same order" method, which would then raise questions "Why this is OrderedDict?".

Comment (araffin) [reply stable_baselines/common/base_class.py]: Even though the ordered is preserved starting at python 3.6, I would find it cleaner to have an ordered dict.
That would also avoid potential bugs if people just want to recover numpy arrays using `param_dict.values()`.

And yes, in fact, `_get_parameter_list` is already kind of public through the keys of the parameters. So, I would put it public. Any thought on that @hill-a ?

File: stable_baselines/common/base_class.py [hunk 3]
```diff
@@ -153,6 +155,49 @@ def _setup_learn(self, seed):
...
+        """
+        # Assume tensorflow graphs are static -> check
+        # that we only call this function once
+        if self._param_load_ops is not None:
+            raise RuntimeError("Parameter load operations have already been created")
+        # For each loadable parameter, create appropiate
+        # placeholder and an assign op, and store them to
+        # self.load_param_ops as dict of variable.name -> (placeholder, assign)
+        loadable_parameters = self._get_parameter_list()
+        # Use OrderedDict to store order for backwards compatability with
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: minor: typo
```suggestion
        # Use OrderedDict to store order for backwards compatibility with
```

File: stable_baselines/common/base_class.py [hunk 4]
```diff
@@ -312,6 +357,51 @@ def action_probability(self, observation, state=None, mask=None, actions=None):
         """
         pass
 
+    def load_parameters(self, load_path_or_dict):
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: as discuss in the issue, I would add an argument (`exact_match=True`) that when not deactivated will raise an error if the dict does not match with the expected one (e.g. missing keys).
This would prevent silent weird behavior.

File: stable_baselines/common/base_class.py [hunk 5]
```diff
@@ -312,6 +357,51 @@ def action_probability(self, observation, state=None, mask=None, actions=None):
...
+            # Assume `load_path_or_dict` is list of ndarrays.
+            # Create param dictionary assuming the parameters are in same order
+            # as _get_parameter_list returns them.
+            params = dict()
+            for i, param_name in enumerate(self._param_load_ops.keys()):
+                params[param_name] = load_path_or_dict[i]
+        else:
+            # Assume a filepath or file-like.
+            # Use existing deserializer to load the parameters
+            _, params = BaseRLModel._load_from_file(load_path)
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: seems like this feature is not tested...
```suggestion
            _, params = BaseRLModel._load_from_file(load_path_or_dict)
```

File: stable_baselines/common/base_class.py [hunk 6]
```diff
@@ -153,6 +155,49 @@ def _setup_learn(self, seed):
...
+        # For each loadable parameter, create appropiate
+        # placeholder and an assign op, and store them to
+        # self.load_param_ops as dict of variable.name -> (placeholder, assign)
+        loadable_parameters = self._get_parameter_list()
+        # Use OrderedDict to store order for backwards compatability with
+        # list-based params
+        self._param_load_ops = OrderedDict()
+        with self.graph.as_default():
+            for param in loadable_parameters:
+                placeholder = tf.placeholder(dtype=param.dtype, shape=param.shape)
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: I don't understand why you need to create those placeholders... I would expect the `param` to suffice.

File: docs/misc/changelog.rst [hunk 7]
```diff
@@ -14,6 +14,9 @@ Release 2.5.2a0 (WIP)
 - The parameter ``filter_size`` of the function ``conv`` in A2C utils now supports passing a list/tuple of two integers (height and width), in order to have non-squared kernel matrix. (@yutingsz)
 - fixed a bug where initial learning rate is logged instead of its placeholder in ``A2C.setup_model`` (@sc420)
 - fixed a bug where number of timesteps is incorrectly updated and logged in ``A2C.learn`` and ``A2C._train_step`` (@sc420)
+- added ``load_parameters`` and ``get_parameters`` for most learning algorithms.
```
Comment (araffin) [inline docs/misc/changelog.rst]: for most? (because of HER?)

File: stable_baselines/common/base_class.py [hunk 8]
```diff
@@ -312,6 +357,51 @@ def action_probability(self, observation, state=None, mask=None, actions=None):
...
+        # Make sure we have assign ops
+        if self._param_load_ops is None:
+            self._setup_load_operations()
+
+        params = None
+        if isinstance(load_path_or_dict, dict):
+            # Assume `load_path_or_dict` is dict of variable.name -> ndarrays we want to load
+            params = load_path_or_dict
+        elif isinstance(load_path_or_dict, list):
+            warnings.warn("Warning: Loading model parameters from a list. This has been replaced " +
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: I would go for a deprecation warning

File: docs/misc/changelog.rst [hunk 9]
```diff
@@ -14,6 +14,9 @@ Release 2.5.2a0 (WIP)
...
+  With these methods, users are able to load and get parameters to/from existing model, without touching tensorflow. (@Miffyli)
+- switched to using dictionaries rather than lists when storing parameters, with tensorflow Variable names being the keys. (@Miffyli)
```
Comment (araffin) [inline docs/misc/changelog.rst]:
```suggestion
- **important change** switched to using dictionaries rather than lists when storing parameters, with tensorflow Variable names being the keys. (@Miffyli)
```

File: stable_baselines/common/base_class.py (hunk 6 above)
Comment (Miffyli) [reply stable_baselines/common/base_class.py]: If you call `param.assign(ndarray)`, it creates a new op in the graph. By doing placeholders and assign ops on them, we avoid constantly creating ops.

File: docs/misc/changelog.rst (hunk 7 above)
Comment (Miffyli) [reply docs/misc/changelog.rst]: Yes, I noticed there were no save/load implemented for HER, so I decided not to touch it (I see it is work-in-progress in other PR).

**Comment by Miffyli** [comment]
I agree on adding some documentation for these functions. I am not sure if it needs to be "advanced users" part, perhaps another example next to load/saves with some additional ES-like-sample in examples section?

**Comment by araffin** [comment]
>perhaps another example next to load/saves with some additional ES-like-sample in examples section?

sounds good ;)

File: stable_baselines/common/base_class.py [hunk 10]
```diff
@@ -312,6 +357,67 @@ def action_probability(self, observation, state=None, mask=None, actions=None):
...
+            feed_dict[placeholder] = param_value
+            # Create list of tf.assign operations for sess.run
+            param_update_ops.append(assign_op)
+            # Keep track which variables are updated
+            not_updated_variables.remove(param_name)
+
+        # Check that we updated all parameters if exact_match=True
+        if exact_match and len(not_updated_variables) > 0:
+            raise RuntimeError("Load dictionary did not contain all variables. " +
+                               "Missing variables: %s" %
```
Comment (araffin) [inline stable_baselines/common/base_class.py]: Please use [format](https://pyformat.info/) instead ;)

**Comment by Miffyli** [comment]
One thing occurred to me that should at least be documented: If you use `load_parameters` to change parameters mid-training, you should also reset optimizer's variables (e.g. momentum). Proper way to do would be to store trainers [slots](https://www.tensorflow.org/api_docs/python/tf/train/Optimizer#slots), select correct trainer variables based on the updated model parameters and calling initializer on them. 

Should this be addressed in this PR, e.g. new function, or leave it as a big red exclamation mark in documentation?

**Comment by araffin** [comment]
>Should this be addressed in this PR, e.g. new function, or leave it as a big red exclamation mark in documentation?

This is related to #301, so for this PR, a warning will be fine. However, contribution are welcomed to solve this issue properly ;)

**Comment by Miffyli** [comment]
Regarding example: ES example is tad silly, given that it does not use the "A2C" part of A2C in any way (just the network and `predict` function). But, with proper handling of trainer variables, this could be nice platform for combining gradient steps and ES ^^.

Regarding OrderedDictionary and `_get_variable_list`: I agree both  on using OrderedDictionary for `get_parameters` and making `_get_variable_list` public (it neatly lists which parameters are important for the model, and advanced users could use these in advantage).

**Comment by araffin** [comment]
>But, with proper handling of trainer variables, this could be nice platform for combining gradient steps and ES

Maybe then you can start training using A2C and then switch to ES to finish optimization (or the other way around). That would make more sense ;)

>Regarding OrderedDictionary and _get_variable_list

Good, I think after that, we are good to merge =)

**Comment by araffin** [comment]
Another thought about the example: you should only grab the param from the policy, otherwise, it will mutate params from the value function too, which does not really makes sense...
